components/supervisor/frontend/src/supervisor-service-client.ts

```diff
@@ -6,24 +6,29 @@
 
 import { WorkspaceInfoResponse } from "@gitpod/supervisor-api-grpc/lib/info_pb";
 
-const checkReady: (kind: 'content' | 'ide' | 'supervisor') => Promise<void> = kind =>
-    fetch(window.location.protocol + '//' + window.location.host + '/_supervisor/v1/status/' + kind + '/wait/true', { credentials: 'include' }).then(response => {
-        if (response.ok) {
-            return;
-        }
-        console.debug(`failed to check whether ${kind} is ready, trying again...`, response.status, response.statusText);
-        return checkReady(kind);
-    }, e => {
-        console.debug(`failed to check whether ${kind} is ready, trying again...`, e);
-        return checkReady(kind);
-    });
-export const supervisorReady = checkReady('supervisor');
-export const ideReady = supervisorReady.then(() => checkReady('ide'));
-export const contentReady = supervisorReady.then(() => checkReady('content'));
+export class SupervisorServiceClient {
+    readonly supervisorReady = this.checkReady('supervisor');
+    readonly ideReady = this.supervisorReady.then(() => this.checkReady('ide'))
+    readonly contentReady = this.supervisorReady.then(() => this.checkReady('content'));
 
-export const fetchWorkspaceInfo = (async () => {
-    await supervisorReady;
-    const response = await fetch(window.location.protocol + '//' + window.location.host + '/_supervisor/v1/info/workspace', { credentials: 'include' });
-    const result = await response.json();
-    return result as WorkspaceInfoResponse.AsObject;
-});
+    async fetchWorkspaceInfo(): Promise<WorkspaceInfoResponse.AsObject> {
+        await this.supervisorReady;
+        const response = await fetch(window.location.protocol + '//' + window.location.host + '/_supervisor/v1/info/workspace', { credentials: 'include' });
+        const result = await response.json();
+        return result as WorkspaceInfoResponse.AsObject;
+    }
+
+    private async checkReady(kind: 'content' | 'ide' | 'supervisor'): Promise<void> {
+        await fetch(window.location.protocol + '//' + window.location.host + '/_supervisor/v1/status/' + kind + '/wait/true', { credentials: 'include' }).then(response => {
+            if (response.ok) {
+                return;
```

Comment:
Should we actually check here that the content is true to reproduce and work around: https://github.com/gitpod-io/gitpod/issues/1939#issuecomment-703248619?

Reply:
yes, we should. If just to make sure this doesn't break if ws-proxy breaks again